File: sediment.m
function  [C1_res, time, x] = sediment(D,w,years)
global dt dx phi N

BC1_top    = 1;  % umol per L
F_bottom  = 0;    % umol per L/y/cm^2
L         = 30; % cm
tend      = years;
C1_init    = 0;


phi =1;
dx = 0.1; % cm
dt = 0.001;% h
x  = 0:dx:L;
N  = size(x,2);

C1_init  = C1_init*ones(N,1);
C1_init(1) = BC1_top;


[AL, AR] = AL_AR_dirichlet(D,w);


C1_old      = C1_init;

time       = 0:dt:tend;
C1_res      = zeros(N,size(time,2));
C1_res(:,1) = C1_init;


for i=2:size(time,2)
    B = AR*C1_old;
    [C1_old, B] = update_bc_dirichlet(C1_old, B, BC1_top, F_bottom, D);
    C1_new      = linalg_solver(AL, B);
    C1_res(:,i) = C1_new;
    C1_old      = C1_new;
    Cold(1)     = BC1_top;
end

%% linalg_solver: x = A \ b
function [x] = linalg_solver(A, b)
    x =  A \ b;

%% update_bc_dirichlet: function description
function [C, B] = update_bc_dirichlet(C, B, BC_top, F_bottom, D)
    global dt dx phi N
    C(1)    = BC_top;
    s       = phi*D*dt/dx/dx;
    B(N) = B(N) + 2*F_bottom*s*dx/phi/D;

%% AL_AR_dirichlet: creates AL and AR matrices with Dirichlet BC
function [AL, AR] = AL_AR_dirichlet(D,w)
    global dt dx phi N
    s       = phi*D*dt/dx/dx;
    q       = phi*w*dt/dx;
    e1      = ones(N,1);
    AL      = spdiags([e1.*(-s/2-q/4) e1.*(1+s) e1.*(-s/2+q/4)],[-1 0 1],N,N);
    AR      = spdiags([e1*(s/2+q/4) e1*(1-s) e1*(s/2-q/4)],[-1 0 1],N,N);
    AL(1,1) = 1; AL(1,2) = 0; AL(N,N) = 1+s; AL(N,N-1) = -s;
    AR(1,1) = 1; AR(1,2) = 0; AR(N,N) = 1-s; AR(N,N-1) = s;

File: sediment.py
    import numpy as np
    import scipy as sp
    from scipy.sparse import spdiags
    from scipy import special
    import matplotlib.pyplot as plt
    # from scikits import umfpack


    def adv_diff(D, w, years, L=30, phi=1, dx=0.1, dt=0.001, bc_top=1):
        """Transport PDE solver"""
        N = int(L / dx) + 1

        [AL, AR] = AL_AR_dirichlet(D, w, phi, dt, dx, N)

        res = np.zeros((N))
        res[0] = bc_top

        yield res

        for _ in range(int(years / dt)):
            res = sp.sparse.linalg.spsolve(AL, AR.dot(res), use_umfpack=True)
            res[0] = bc_top
            yield res


    def AL_AR_dirichlet(D, w, phi, dt, dx, N):
        """AL_AR_dirichlet: creates AL and AR matrices with Dirichlet BC"""
        s = phi * D * dt / dx / dx  #
        q = phi * w * dt / dx  #
        e1 = np.ones((N, 1))  #
        AL = spdiags(np.concatenate((e1 * (-s / 2 - q / 4), e1 * (1 + s), e1 * (-s / 2 + q / 4)), axis=1).T, [-1, 0, 1], N, N, format='csc')
        AR = spdiags(np.concatenate((e1 * (s / 2 + q / 4), e1 * (1 - s), e1 * (s / 2 - q / 4)), axis=1).T, [-1, 0, 1], N, N, format='csc')
        AL[0, :2] = 1, 0
        AL[-1, -2:] = -s, 1 + s
        AR[0, :2] = 1, 0
        AR[-1, -2:] = s, 1 - s
        return AL, AR


    if __name__ == '__main__':
        #
        D = 0.5
        w = 3
        t = 10

        C = np.array(list(adv_diff(D, w, t)))

    # L = 30
    # dx = 0.1
    # x = np.linspace(0, L, L / dx + 1)
    # sol = 1 / 2 * (special.erfc((x - w * t) / 2 / np.sqrt(D * t)) + np.exp(w * x / D) * special.erfc((x + w * t) / 2 / np.sqrt(D * t)))

    # plt.plot(x, C[:,-1])
    # plt.plot(x, sol)
    # plt.show()
